### PR TITLE
Improvements to table: `restyle` fix, 0 row/column and misc. fixes; jasmine tests

### PR DESCRIPTION
**Pull request by monfera**
- `restyle` fix was urgent due to https://github.com/plotly/plotly.js/issues/2092
- why not take the opportunity to add jasmine coverage for most of what exists
- tests uncovered misc. special case issues which were fixed

### CODE REVIEW

**Comment by VeraZab** [comment]
I'll test out this branch on streambed, see if the restyle calls I need work now, thanks!

**Comment by VeraZab** [comment]
ok, I see the improvements, 

here's a few observations: 
1) there's  inconsistency where I I need to wrap my data into an extra set of arrays to make some calls work:
for example: https://codepen.io/veraz/pen/EwMLBQ?editors=1010
I can make Plotly.restyle('graphDiv', {'header.values': [[['A'], ['B'], ['C']]]}) work, with one extra set of arrays. It's the same for a few other calls like: Plotly.restyle(gd, {'cells.font.size': [[16, 5]]})

2) A restyle to `cells.fill.color` behaves unexpectedly, it should color by row right? not by column..

I'd really like to see that extra [] wrapping fixed in this pr, would really help for the workspace integration.

Thanks!

File: test/jasmine/tests/table_test.js [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            }
+
+            restyleValues('cells', 0, [['new cell content 1', 'new cell content 2']])()
+                .then(restyleValues('cells', 2, [[0, 0.1]]))
+                .then(restyleValues('header', 1, [['Species top', 'Species bottom']]))
+                .then(done);
+        });
+
+        it('Calling `Plotly.relayout` with string should amend the preexisting parcoords', function(done) {
+            expect(gd.layout.width).toEqual(1000);
```
Comment (VeraZab) [inline test/jasmine/tests/table_test.js]: preexisting 'tables' instead of 'parcoords'

File: test/jasmine/tests/table_test.js [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            Plotly.relayout(gd, 'width', 500).then(function() {
+                expect(gd.data.length).toEqual(1);
+                expect(gd.layout.width).toEqual(500);
+                done();
+            });
+        });
+
+        it('Calling `Plotly.relayout` with object should amend the preexisting parcoords', function(done) {
+            expect(gd.layout.width).toEqual(1000);
```
Comment (VeraZab) [inline test/jasmine/tests/table_test.js]: can we add a few tests testing some of those array attributes?

**Comment by monfera** [comment]
Hi @VeraZab yes #1 is the way `plotly.js` works when it's an array attribute of traces (of which there can be multiple). Re #2, it can do per column, per row, per cell or combination. The idea is that if the columns in a three-column table are `[A, B, C]`, then eg. `B` can be a scalar value or color; but it can also be an array. So you can have 3 columns, 2 rows styled as `[[A1, A2], [B1, B2], [C1, C2]]` on a per cell basis, or you can do just `[[A1, A2]]` ie. leave other columns undefined, in which case they inherit colors, so it's basically per-row styling. I'll make an example.

**Comment by monfera** [comment]
Per column: as you did above
Per row: https://codepen.io/anon/pen/vePreY?editors=1010
Combined: https://codepen.io/anon/pen/XeGYzB?editors=1010

**Comment by monfera** [comment]
@VeraZab this version replaces cell contents in one call, ie. not of just one column at a time: https://codepen.io/anon/pen/dVrKqb?editors=1010 - I'm adding jasmine test cases for such things, and what you're pointing out above as well.

**Comment by VeraZab** [comment]
oh for #1, yeah sorry there's a bunch of things that I'm realizing are not working as expected in the workspace, so I was making the calls from the workspace console to test your pr, rather from the workspace ui, which already has those conventions pre-built, yeah I think that's fine.

> @VeraZab this version replaces cell contents in one call, ie. not of just one column at a time: https://codepen.io/anon/pen/dVrKqb?editors=1010 - I'm adding jasmine test cases for such things, and what you're pointing out above as well.

yeah, both are supported right, from what I've seen in my tests, and that's great, should be that way, and yes, let's add some tests

Ok, well, aside from a few extra tests, I'm good with this pr 💃

**Comment by etpinard** [comment]
Oh shit, what Vera points out will be very tricky to fix. We can't rely on what we currently do for other 2d array trace types as table is column major.

**Comment by VeraZab** [comment]
@etpinard I think I've made a mistake in that comment, about array wrapping.

I've just checked for a scatter trace, and it's true that when I want to send an update to `marker.color`, I don't just do Plotly.restyle(gd, {'marker.color': [1, 2, 3]})

I wrap the [1, 2, 3] array into another array and the correct call is
Plotly.restyle(gd, {'marker.color': [[1, 2, 3]]})

so my bad, I think this pr's good to go, after tests

**Pull request merged**